solutions/devpsrint-michael-douglas-1/GitHubApp/Service/Service.swift

```diff
@@ -7,10 +7,42 @@
 
 import Foundation
 
-struct Service {
-
-    func fetchList(_ completion: ([String]) -> Void) {
+protocol ServiceProtocol {
+    
+    func fetchRepositoryList(_ userName: String, _ completion: @escaping ([Repository]?) -> Void)
+}
 
-        completion(["Repository 1", "Repository 2", "Repository 3"])
+struct Service: ServiceProtocol {
+    
+    func fetchRepositoryList(_ userName: String, _ completion: @escaping ([Repository]?) -> Void) {
+        
+        let url = URL(string: "https://api.github.com/users/\(userName)/repos")!
+        self.performRequest(url, completion: completion)
+    }
+    
+    func performRequest<T: Decodable>(_ url: URL, completion: @escaping (T?) -> Void) {
+        
+        let dataTask = URLSession.shared.dataTask(with: url) { data, response, error in
+            
+            if let _ = error {
+                completion(nil)
+                return
+            }
+            
+            if let data = data {
+                
+                let jsonDecodable = JSONDecoder()
+                jsonDecodable.keyDecodingStrategy = .convertFromSnakeCase
+                do {
+                    let repositories = try jsonDecodable.decode(T.self, from: data)
+                    completion(repositories)
+                } catch {
+                    print(error)
+                    completion(nil)
+                }
+            }
+        }
+        dataTask.resume()
```

Comment:
Aqui você pode usar o `guard let` ao invés de `if let`, tornando seu código mais clean. Segue um material bacana pra tu estudar para entender a diferença entre eles e quando usar cada um: https://abhimuralidharan.medium.com/if-let-if-var-guard-let-and-defer-statements-in-swift-4f87fe857eb6

```suggestion
        let dataTask = URLSession.shared.dataTask(with: url) { data, response, error in
            
            guard let data = data else { 
                completion(nil)
                return
            }

            let jsonDecodable = JSONDecoder()
            jsonDecodable.keyDecodingStrategy = .convertFromSnakeCase
            
            do {
                 let decodedObject = try jsonDecodable.decode(T.self, from: data)
                 completion(decodedObject)
            } catch {
                 print(error)
                 completion(nil)
            }
        }.resume()
```